File: 50dias/python/46.py
import random

def LeTextoDoArquivo(nomeDoArquivo):
    input_file = open(nomeDoArquivo, 'r')
    texto = input_file.read()
    input_file.close()
    return texto

def EscreveTextoNoArquivo(texto, nomeDoArquivo):
    output_file = open(nomeDoArquivo, 'w')
    output_file.write(texto)
    output_file.close()


def EmbaralhaPalavraMantendoPrimeiraUltima(palavra):
    tamanhoPalavra = len(palavra)
    novaPalavra = palavra
    if tamanhoPalavra > 1:
        meio = palavra[1:tamanhoPalavra-1]
        novoMeio = ''.join(random.sample(meio, len(meio)))
        novaPalavra = palavra[0] + novoMeio + palavra[tamanhoPalavra-1]
    return novaPalavra

def main():
    texto = LeTextoDoArquivo('entrada.txt')
    novoTexto = ""
    for palavra in texto.split(' '):
        novoTexto += EmbaralhaPalavraMantendoPrimeiraUltima(palavra) + ' '

    EscreveTextoNoArquivo(novoTexto, 'saida.txt')

main()

File: 50dias/cpp/46.cpp
#include <iostream>
#include <random>
#include <fstream>
#include <list>
#include <string>
#include <sstream>

using namespace std;

void LerArquivo(
    const string& nomeDoArquivo,
    list<string>& texto)
{
    ifstream inputFile;
    inputFile.open(nomeDoArquivo);

    string linha;
    while (getline(inputFile, linha)) {
        texto.push_back(linha);
    }
    inputFile.close();
}

void EscreveArquivo(
    const string& nomeDoArquivo,
    const list<string> texto
)
{
    ofstream outputFile;
    outputFile.open(nomeDoArquivo);

    for (auto linha : texto) {
        outputFile << linha << endl;
    }
    outputFile.close();
}

void EmbaralhaPalavraMantendoPrimeiraUltimaLetra(
    string& palavra
)
{
    if (palavra.size() > 1) {
        std::random_device rd;
        std::mt19937 g(rd());

        std::shuffle(++palavra.begin(),
            --palavra.end(),
            g);
    }
}

int main()
{
    list<string> texto;
    LerArquivo("entrada.txt", texto);

    list<string>novoTexto;
    for (auto linha : texto) {
        stringstream ss(linha);
        string palavra;
        string novaLinha;
        while (ss >> palavra) {
            EmbaralhaPalavraMantendoPrimeiraUltimaLetra(palavra);
            novaLinha += palavra + " ";
        }
        novoTexto.push_back(novaLinha);
    }

    EscreveArquivo("saida.txt", novoTexto);
    return 0;
}